text/0000-allow-sharing-of-LWC-across-namespaces.md

```diff
@@ -0,0 +1,230 @@
+
+For now we will honor isExposed flag and display error for in valid scenarios listed below. However, in the future we will completely rely on namespaceScope tag. 
+
+**Valid Cases**
+
+| namespaceScope    | isExposed       | Comments                                                      |
+|----------|-----------------|---------------------------------------------------------------|
+| Global   | TRUE            | Open to the world                                             |
+| Targeted | FALSE           | Open to the namespaces listed in the NamespaceBundle.xml file |
+| Local    | FALSE           | Open to local name space only                                 |
+| MISSING  | TRUE            | Open to the world                                             |
+| MISSING  | FALSE           | Open to local name space only                                 |
+| MISSING  | MISSING (FALSE) | Open to local name space only                                 |
+
+**Error Cases**
+
+| namespaceScope    | isExposed | Comments                   | Errors             |
+|----------|-----------|----------------------------|--------------------|
+| Global   | FALSE     | Failed during compile time | Compile time error |
+| Targeted | TRUE      | Failed during compile time | Compile time error |
+| Local    | TRUE      | Failed during compile time | Compile time error |
+
+
+```
+<!-- LightingComponentBundle.xml -->
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>49.0</apiVersion>
+    <isExposed>false</isExposed>
+    <namespaceScope>global/targeted/local</namespaceScope> 
+</LightningComponentBundle>
+```
+```
+<!-- NamespaceBundle.xml -->
+<?xml version="1.0" encoding="UTF-8"?>
+<NamespaceBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <namespaces>
+        <namespace>force</namespace>
+        <namespace>forcechatter</namespace>
+        <namespace>ui</namespace>
+    </namespaces>
+</NamespaceBundle>
+```
+
+Pros
+
+* Honor isExposed flag.
+* Sharing is easy as you have to edit one single file to add or remove access to the namespaces.
+
+Cons
+
+* Individual component cannot be expose to different namespaces. It is controlled at namespace level so either all or none.
+* Learning curve as we are introducing new flag.
+* Complex implementation as we have to set precedence logic in the code.
+
+## High level design
+
+In order to honor the sharing of LWC across namespaces, we will be doing validation at difference stages viz.
+
+#### Compile Time Validation : 
+
+During compile time all we do is compile the component either using RINO (Aura) or NodeJS (LWC) compiler and load it into registry (BundleAwareDefRegistry). Since compiler has no context of namespace there is nothing much we can do here. During compile time we can validate Bundle xml file and error out as listed in the table above. 
+

```

Comment:
I would consider LAS the experts, but in my view LWC on core is tied into the Aura framework for compiling. In Aura, compiling also involves linking. Every time you compile a component it's also gets linked. In Aura look at AccessChecker.computeModuleAccess. That's where this enforcement should happen.

Reply:
Yes we are planning to do the linking time access for imports. However, with the component, if users have mis configured say isExposed to true but namespaceScope is local in that case we wanted to throw QFE.